File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/e69b41e64e0736af6aa94f4dd3062280.go
package main

import (
	"fmt"
	"math"
)

const MOD = 1e9 + 7

type TreeSegment struct {
	S       []Node
	n       int
	magic   func(int, int) int
	applyMod func(int, int, int, int) int
	mergeMods func(int, int, int, int) int
}

type Node struct {
	haveMod bool
	l, r    int
	value   int
	mod     int
}

func NewTreeSegment(src []int, magic func(int, int) int, applyMod func(int, int, int, int) int, mergeMods func(int, int, int, int) int) *TreeSegment {
	ts := &TreeSegment{
		magic:   magic,
		applyMod: applyMod,
		mergeMods: mergeMods,
	}
	ts.init(src)
	return ts
}

func (ts *TreeSegment) init(src []int) {
	ts.n = len(src)
	ts.S = make([]Node, 4*ts.n)
	ts.build(0, 0, ts.n, src)
}

func (ts *TreeSegment) build(i, tl, tr int, src []int) {
	if tl+1 == tr {
		ts.S[i].value = src[tl]
	} else {
		mid := (tl + tr) / 2
		ts.S[i].l = 2*i + 1
		ts.S[i].r = 2*i + 2
		ts.build(ts.S[i].l, tl, mid, src)
		ts.build(ts.S[i].r, mid, tr, src)
		ts.recalcValue(i, tl, tr)
	}
}

func (ts *TreeSegment) recalcValue(i, tl, tr int) {
	if tl+1 != tr {
		mid := (tl + tr) / 2
		ts.S[i].value = ts.magic(ts.getValue(ts.S[i].l, tl, mid), ts.getValue(ts.S[i].r, mid, tr))
	}
}

func (ts *TreeSegment) getValue(i, tl, tr int) int {
	if !ts.S[i].haveMod {
		return ts.S[i].value
	}
	return ts.applyMod(ts.S[i].value, ts.S[i].mod, tl, tr)
}

func (ts *TreeSegment) update(ql, qr, mod int) {
	ts.updateRec(0, 0, ts.n, ql, qr, mod)
}

func (ts *TreeSegment) updateRec(i, tl, tr, ql, qr, mod int) {
	if ql <= tl && tr <= qr {
		ts.addMod(i, tl, tr, mod)
	} else {
		mid := (tl + tr) / 2
		if ql < mid {
			ts.updateRec(ts.S[i].l, tl, mid, ql, qr, mod)
		}
		if mid < qr {
			ts.updateRec(ts.S[i].r, mid, tr, ql, qr, mod)
		}
		ts.recalcValue(i, tl, tr)
	}
}

func (ts *TreeSegment) addMod(i, tl, tr, sub int) {
	if ts.S[i].haveMod {
		ts.S[i].mod = ts.mergeMods(ts.S[i].mod, sub, tl, tr)
	} else {
		ts.S[i].mod = sub
		ts.S[i].haveMod = true
	}
}

func (ts *TreeSegment) query(ql, qr int) int {
	return ts.queryRec(0, 0, ts.n, ql, qr)
}

func (ts *TreeSegment) queryRec(i, tl, tr, ql, qr int) int {
	if ql <= tl && tr <= qr {
		return ts.S[i].value
	}
	mid := (tl + tr) / 2
	res := 0
	if ql < mid {
		res = ts.queryRec(ts.S[i].l, tl, mid, ql, qr)
	}
	if mid < qr {
		res = ts.magic(res, ts.queryRec(ts.S[i].r, mid, tr, ql, qr))
	}
	return res
}

func floorMod(a, b int) int {
	if a%b == 0 {
		return 0
	}
	if a >= 0 && b >= 0 {
		return a % b
	}
	if a <= 0 && b <= 0 {
		return a % b
	}
	return int(math.Abs(float64(b))) - (int(math.Abs(float64(a))) % int(math.Abs(float64(b))))
}

func main() {
	var n, q int
	fmt.Scan(&n, &q)

	pos := make([]int, n)
	w := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&pos[i])
	}
	for i := 0; i < n; i++ {
		fmt.Scan(&w[i])
	}

	ts := NewTreeSegment(w, func(a, b int) int {
		return a + b
	}, func(value, mod, _, _ int) int {
		return mod
	}, func(old, new, _, _ int) int {
		return new
	})

	subw := make([]int, n)
	for i := 0; i < n; i++ {
		subw[i] = (w[i] * (i - pos[i])) % MOD
	}

	resp := NewTreeSegment(subw, func(a, b int) int {
		return (a + b) % MOD
	}, func(value, mod, _, _ int) int {
		return mod
	}, func(old, new, _, _ int) int {
		return new
	})

	for i := 0; i < q; i++ {
		var x, y int
		fmt.Scan(&x, &y)
		if x < 0 {
			ind := -x - 1
			newVal := y
			ts.update(ind, ind+1, newVal)
			resp.update(ind, ind+1, (newVal*(ind-pos[ind]))%MOD)
		} else {
			l := x - 1
			r := y

			sm := ts.query(l, r)

			bestL := ts.findRight(l, func(x int) bool {
				return x*2 >= sm
			}, 0)

			x := pos[bestL] - bestL + l

			left := 0
			if l < bestL {
				left = (resp.query(l, bestL) + ts.query(l, bestL)*(x-l)) % MOD
			}
			right := 0
			if bestL < r {
				right = (resp.query(bestL, r) + ts.query(bestL, r)*(x-l)) % MOD
			}
			fmt.Println(floorMod(left-right, MOD))
		}
	}
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/e69b41e64e0736af6aa94f4dd3062280.java
import java.util.*;
import java.io.*;

public class e69b41e64e0736af6aa94f4dd3062280_nc {
    static final long MOD = 1000000007;

    static class TreeSegment<Value, Mod> {
        static class Node<Value, Mod> {
            boolean haveMod = false;
            int l = -1, r = -1;
            Value value;
            Mod mod;
        }

        private final List<Node<Value, Mod>> S = new ArrayList<>();
        private int n;
        private final BiFunction<Value, Value, Value> magic;
        private final BiFunction<Value, Mod, Value> applyMod;
        private final TriFunction<Mod, Mod, Mod> mergeMods;

        public TreeSegment(List<Value> src, BiFunction<Value, Value, Value> magic,
                           BiFunction<Value, Mod, Value> applyMod,
                           TriFunction<Mod, Mod, Mod> mergeMods) {
            this.magic = magic;
            this.applyMod = applyMod;
            this.mergeMods = mergeMods;
            init(src);
        }

        private void init(List<Value> src) {
            n = src.size();
            init(0, n, src);
        }

        private int init(int tl, int tr, List<Value> src) {
            int i = createNode();
            if (tl + 1 == tr) {
                S.get(i).value = src.get(tl);
            } else {
                int mid = (tl + tr) / 2;
                S.get(i).l = init(tl, mid, src);
                S.get(i).r = init(mid, tr, src);
                recalcValue(i, tl, tr);
            }
            return i;
        }

        private int createNode() {
            S.add(new Node<>());
            return S.size() - 1;
        }

        private void recalcValue(int i, int tl, int tr) {
            if (tl + 1 != tr) {
                int mid = (tl + tr) / 2;
                S.get(i).value = magic.apply(getValue(S.get(i).l, tl, mid), getValue(S.get(i).r, mid, tr));
            }
        }

        private Value getValue(int i, int tl, int tr) {
            if (!S.get(i).haveMod) return S.get(i).value;
            return applyMod.apply(S.get(i).value, S.get(i).mod);
        }

        public Value query(int ql, int qr) {
            return query(0, 0, n, ql, qr);
        }

        private Value query(int i, int tl, int tr, int ql, int qr) {
            push(i, tl, tr);
            if (ql <= tl && tr <= qr) {
                return S.get(i).value;
            } else {
                int mid = (tl + tr) / 2;
                if (intersects(tl, mid, ql, qr) && intersects(mid, tr, ql, qr)) {
                    return magic.apply(query(S.get(i).l, tl, mid, ql, qr), query(S.get(i).r, mid, tr, ql, qr));
                } else if (intersects(tl, mid, ql, qr)) {
                    return query(S.get(i).l, tl, mid, ql, qr);
                } else if (intersects(mid, tr, ql, qr)) {
                    return query(S.get(i).r, mid, tr, ql, qr);
                } else {
                    throw new AssertionError();
                }
            }
        }

        private void push(int i, int tl, int tr) {
            if (S.get(i).haveMod) {
                S.get(i).value = applyMod.apply(S.get(i).value, S.get(i).mod);
                if (tl + 1 != tr) {
                    int mid = (tl + tr) / 2;
                    addMod(S.get(i).l, tl, mid, S.get(i).mod);
                    addMod(S.get(i).r, mid, tr, S.get(i).mod);
                }
                S.get(i).haveMod = false;
            }
        }

        private void addMod(int i, int tl, int tr, Mod sub) {
            if (S.get(i).haveMod) {
                S.get(i).mod = mergeMods.apply(S.get(i).mod, sub);
            } else {
                S.get(i).mod = sub;
                S.get(i).haveMod = true;
            }
        }

        private boolean intersects(int tl, int tr, int ql, int qr) {
            return !(tr <= ql || qr <= tl);
        }

        public void update(int ql, int qr, Mod mod) {
            update(0, 0, n, ql, qr, mod);
        }

        private void update(int i, int tl, int tr, int ql, int qr, Mod mod) {
            push(i, tl, tr);
            if (ql <= tl && tr <= qr) {
                addMod(i, tl, tr, mod);
            } else {
                int mid = (tl + tr) / 2;
                if (intersects(tl, mid, ql, qr)) {
                    update(S.get(i).l, tl, mid, ql, qr, mod);
                }
                if (intersects(mid, tr, ql, qr)) {
                    update(S.get(i).r, mid, tr, ql, qr, mod);
                }
                recalcValue(i, tl, tr);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long n = Long.parseLong(st.nextToken());
        long q = Long.parseLong(st.nextToken());

        List<Integer> pos = new ArrayList<>();
        List<Integer> w = new ArrayList<>();
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            pos.add(Integer.parseInt(st.nextToken()));
        }
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            w.add(Integer.parseInt(st.nextToken()));
        }

        TreeSegment<Integer, Integer> ts = new TreeSegment<>(w,
                Integer::sum,
                (value, mod) -> mod,
                (old, _new) -> _new);

        List<Integer> subw = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            subw.add((int) ((w.get(i) * (i - pos.get(i))) % MOD));
        }

        TreeSegment<Integer, Integer> resp = new TreeSegment<>(subw,
                (a, b) -> (a + b) % MOD,
                (value, mod) -> mod,
                (old, _new) -> _new);

        for (int i = 0; i < q; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            if (x < 0) {
                int ind = -x - 1;
                int _new = y;
                ts.update(ind, ind + 1, _new);
                resp.update(ind, ind + 1, (_new * (ind - pos.get(ind))) % MOD);
            } else {
                int l = x - 1;
                int r = y;

                int sm = ts.query(l, r);

                int bestL = ts.findRight(l, xValue -> xValue * 2 >= sm, 0);

                int xValue = pos.get(bestL) - bestL + l;

                int left = 0;
                if (l < bestL) {
                    left = (resp.query(l, bestL) + ts.query(l, bestL) * (xValue - l)) % MOD;
                }
                int right = 0;
                if (bestL < r) {
                    right = (resp.query(bestL, r) + ts.query(bestL, r) * (xValue - l)) % MOD;
                }
                System.out.println(floorMod(left - right, MOD));
            }
        }
    }

    private static long floorMod(long a, long b) {
        if (a % b == 0) return 0;
        if (a >= 0 && b >= 0) return a % b;
        if (a <= 0 && b <= 0) return a % b;
        return Math.abs(b) - (Math.abs(a) % Math.abs(b));
    }

    @FunctionalInterface
    interface TriFunction<A, B, C> {
        C apply(A a, B b);
    }
}
